packages/core/src/lib/light_push/light_push.ts

```diff
@@ -16,61 +16,123 @@ import { pipe } from "it-pipe";
 import { Uint8ArrayList } from "uint8arraylist";
 
 import { StreamManager } from "../stream_manager/index.js";
+import { selectOpenConnection } from "../stream_manager/utils.js";
 
-import { PushRpc } from "./push_rpc.js";
+import { PushRpcV2 } from "./push_rpc.js";
+import { PushRpcV3 } from "./push_rpc_v3.js";
+import { isSuccess as isV3Success } from "./utils.js";
 import { isRLNResponseError } from "./utils.js";
 
 const log = new Logger("light-push");
 
-export const LightPushCodec = "/vac/waku/lightpush/2.0.0-beta1";
+export const LightPushCodecV2 = "/vac/waku/lightpush/2.0.0-beta1";
+export const LightPushCodecV3 = "/vac/waku/lightpush/3.0.0";
+export const LightPushCodecs = [LightPushCodecV3, LightPushCodecV2];
 export { PushResponse };
 
-type PreparePushMessageResult = ThisOrThat<"query", PushRpc>;
+type PreparePushMessageResult = ThisOrThat<
+  "query",
+  PushRpcV2 | PushRpcV3,
+  "error",
+  LightPushError
+>;
 
 /**
  * Implements the [Waku v2 Light Push protocol](https://rfc.vac.dev/spec/19/).
  */
 export class LightPushCore {
   private readonly streamManager: StreamManager;
 
-  public readonly multicodec = LightPushCodec;
+  public readonly multicodec = LightPushCodecV2;
+  public readonly multicodecs = LightPushCodecs;
 
-  public constructor(libp2p: Libp2p) {
-    this.streamManager = new StreamManager(LightPushCodec, libp2p.components);
+  public constructor(private libp2p: Libp2p) {
+    this.streamManager = new StreamManager(LightPushCodecV2, libp2p.components);
+  }
+
+  private async getProtocolStream(
+    peerId: PeerId
+  ): Promise<{ stream: Stream; protocol: string }> {
+    const peer = await this.libp2p.peerStore.get(peerId);
+    const protocols = peer.protocols;
+
+    const supportsV3 = protocols.includes(LightPushCodecV3);
+    const supportsV2 = protocols.includes(LightPushCodecV2);
+
+    if (!supportsV2 && !supportsV3) {
+      throw new Error("Peer does not support any Light Push protocol");
+    }
+
+    // TODO: Remove forced v2 fallback - currently disabled to test v3 protocol communication
+    // Prefer v3 protocol when available, fallback to v2
+    const protocol = supportsV3 ? LightPushCodecV3 : LightPushCodecV2;
+
+    let stream: Stream;
+    try {
+      const connections = this.libp2p.getConnections(peerId);
+      const connection = selectOpenConnection(connections);
+
+      if (!connection) {
+        throw new Error("No open connection to peer");
+      }
+
+      stream = await connection.newStream(protocol);
+    } catch (error) {
+      if (supportsV3 && supportsV2) {
+        // TODO: This fallback should be evaluated for v3 compatibility
+        log.warn(
+          `Failed to create v3 stream, falling back to v2. Error: ${error}`
+        );
+        log.info("Protocol negotiation chose v2 due to failure in v3");
+        stream = await this.streamManager.getStream(peerId);
```

Comment:
Why using `streamManager` when it has fails but `stream = await connection.newStream(protocol)` on the first go?